services/gundeck/src/Gundeck/API/Public.hs

```diff
@@ -228,28 +99,45 @@ listTokensH (uid ::: _) =
 --
 -- (arianvp): I am not sure why it is convenient for clients to distinct
 -- between these two cases.
-paginateH :: JSON ::: UserId ::: Maybe ByteString ::: Maybe ClientId ::: Range 100 10000 Int32 -> Gundeck Response
-paginateH (_ ::: uid ::: sinceRaw ::: clt ::: size) = do
-  Notification.PaginateResult gap page <- Notification.paginate uid (join since) clt size
-  pure . updStatus gap . json $ (page :: Public.QueuedNotificationList)
+paginateUntilV2 ::
+  UserId ->
+  Maybe Public.RawNotificationId ->
+  Maybe ClientId ->
+  Maybe (Range 100 10000 Int32) ->
+  Gundeck Public.GetNotificationsResponse
+paginateUntilV2 uid mbSince mbClient mbSize = do
+  let size = fromMaybe (unsafeRange 1000) mbSize
+  Notification.PaginateResult gap page <- Notification.paginate uid (join since) mbClient size
+  pure $
+    if gap
+      then Public.GetNotificationsWithStatusNotFound page
+      else case since of
+        Just (Just _) -> Public.GetNotificationsSuccess page
+        Nothing -> Public.GetNotificationsSuccess page
+        Just Nothing -> Public.GetNotificationsWithStatusNotFound page
   where
-    since :: Maybe (Maybe NotificationId)
-    since = parseUUID <$> sinceRaw
-    parseUUID :: ByteString -> Maybe NotificationId
-    parseUUID = UUID.fromASCIIBytes >=> isV1UUID >=> pure . Id
+    since :: Maybe (Maybe Public.NotificationId)
+    since = parseUUID <$> mbSince
+
+    parseUUID :: Public.RawNotificationId -> Maybe Public.NotificationId
+    parseUUID = (UUID.fromASCIIBytes . Public.unRawNotificationId) >=> isV1UUID >=> pure . Id
+
     isV1UUID :: UUID -> Maybe UUID
     isV1UUID u = if UUID.version u == 1 then Just u else Nothing
-    updStatus :: Bool -> Response -> Response
-    updStatus True = setStatus status404
-    updStatus False = case since of
-      Just (Just _) -> id
-      Nothing -> id
-      Just Nothing -> setStatus status404
-
-getByIdH :: JSON ::: UserId ::: NotificationId ::: Maybe ClientId -> Gundeck Response
-getByIdH (_ ::: uid ::: nid ::: cid) =
-  json @Public.QueuedNotification <$> Notification.getById uid nid cid
 
-getLastH :: JSON ::: UserId ::: Maybe ClientId -> Gundeck Response
-getLastH (_ ::: uid ::: cid) =
-  json @Public.QueuedNotification <$> Notification.getLast uid cid
+paginate ::
+  UserId ->
+  Maybe Public.NotificationId ->
+  Maybe ClientId ->
+  Maybe (Range 100 10000 Int32) ->
+  Gundeck (Maybe Public.QueuedNotificationList)
+paginate uid mbSince mbClient mbSize = do
```

Comment:
is this a copy of [services/gundeck/src/Gundeck/Notification.hs](https://github.com/wireapp/wire-server/pull/2769/files#diff-fe111b42043c223d14ca226285882b97eda97c57323b019b3a63423ad8308d64)?  do you want to remove the latter, than?

Reply:
I don't understand this question. The link seems to point to something unrelated.